go-controller/pkg/clustermanager/persistentips/allocator.go

```diff
@@ -0,0 +1,94 @@
+package persistentips
+
+import (
+	"fmt"
+	"net"
+	"strings"
+
+	"k8s.io/klog/v2"
+
+	ipamclaimsapi "github.com/k8snetworkplumbingwg/ipamclaims/pkg/crd/ipamclaims/v1alpha1"
+
+	ipam "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/allocator/ip"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/allocator/ip/subnet"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+// Allocator acts on IPAMClaim events handed off by the cluster network
+// controller and allocates or releases IPs for IPAMClaims.
+type Allocator struct {
+	kube kube.InterfaceOVN
+
+	// ipAllocator of IPs within subnets
+	ipAllocator subnet.NamedAllocator
+}
+
+// NewPersistentIPsAllocator builds a new PersistentIPsAllocator
+func NewPersistentIPsAllocator(kube kube.InterfaceOVN, ipAllocator subnet.NamedAllocator) *Allocator {
+	pipsAllocator := &Allocator{
+		kube:        kube,
+		ipAllocator: ipAllocator,
+	}
+
+	return pipsAllocator
+}
+
+// Reconcile updates an IPAMClaim with the IP addresses allocated to the pod's
+// interface
+func (a *Allocator) Reconcile(ipamClaim *ipamclaimsapi.IPAMClaim, ips []string) error {
+	klog.V(5).Infof("Reconciling IPAMLease %q", ipamClaim.Name)
+	if len(ipamClaim.Status.IPs) > 0 {
+		klog.V(5).Infof("Already have neat lookin' IPs for: %q. Bail out !", ipamClaim.Name)
+		return nil
+	}
+
+	if err := a.kube.UpdateIPAMLeaseIPs(ipamClaim, ips); err != nil {
+		return fmt.Errorf(
+			"failed to update the allocation %q with allocations %q: %v",
+			ipamClaim.Name,
+			strings.Join(ips, ","),
+			err,
+		)
+	}
+
+	return nil
+}
+
+// Sync initializes the IPs allocator with the IPAMClaims already existing on
+// the cluster. For live pods, therse are already allocated, so no error will
+// be thrown (e.g. we ignore the `ipam.IsErrAllocated` error
+func (a *Allocator) Sync(objs []interface{}) error {
+	var ips []*net.IPNet
+	for _, obj := range objs {
+		pips, ok := obj.(*ipamclaimsapi.IPAMClaim)
+		if !ok {
+			klog.Errorf("Could not cast %T object to *ipamclaimsapi.IPAMClaim", obj)
+			continue
+		}
+		ipnets, err := util.ParseIPNets(pips.Status.IPs)
+		if err != nil {
+			return fmt.Errorf("failed at parsing IP when allocating persistent IPs: %v", err)
+		}
+		ips = append(ips, ipnets...)
+	}
+	if len(ips) > 0 {
+		if err := a.ipAllocator.AllocateIPs(ips); err != nil && !ipam.IsErrAllocated(err) {
```

Comment:
do we rollback `AllocateIPs` if claim status update failed ? (i.e if claim was deleted meanwhile)